### ISSUE TITLE
Typescript 5 ES Decorators can overwrite fields that are modified by prototype

### ISSUE THREAD

**Issue by cheit-epic** [microsoft/TypeScript#54264]
# Bug Report

When `strictPropertyInitialization=false`, if you have a class that uses a decorator somewhere, and the prototype is later modified to tweak the first field in the class, then the field becomes `undefined` instead of what the prototype says.

### 🔎 Search Terms

decorator, undefined

### 🕗 Version & Regression Information

- I was unable to test this on prior versions because non-experimental decorators are only available in 5.0+.

### ⏯ Playground Link

[Playground link with relevant code][(https://www.typescriptlang.org/play?#code/PTAEFkE9QYwewCYFNQHM5IM6gBZIE5JA)](https://www.typescriptlang.org/play?strictPropertyInitialization=false#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVwCbBrQCGM0AFDMVAObAwBccxKAngDRxEoyiPNsAlHADeAXwBQEtABtiAZ3lwAYhAiiJcRGoBG1JiiQBbHcCgBuKVoACBIlFLRNcMEh0zMaOHqjlh4iUkJAHkdACtCGAA6BXlMGhRyVQgosCgIMhhWMGBOEW0IHyYAJgAGODFBS255CBlgKJkIGnIUYAB3FTU-KIRdaiqJIA)

### 💻 Code

```ts
export function decorator(target: any, context: any) {}

class Foo {
  foobar: number;

  @decorator
  public bar() {}
}

Object.assign(Foo.prototype, { foobar: 20 });
console.log(new Foo().foobar);

```

### 🙁 Actual behavior

Prints `undefined`.
This is because the emitted JS sees this field as uninitialized, so sets it to `void 0`, which happens after the prototype was changed.
```ts
constructor() {
  this.foobar = (__runInitializers(this, _instanceExtraInitializers), void 0);
}
```

### 🙂 Expected behavior

Prints `20`

**Comment by fatcerberus** [comment]
This is the expected behavior; it’s what this code would do in vanilla JS with native decorators.

**Comment by cheit-epic** [comment]
> This is the expected behavior; it’s what this code would do in vanilla JS with native decorators.

Makes perfect sense. It would still be really nice to have some documentation describing behavior changes between experimental decorators and the new ones to make it easier to anticipate issues when migrating.

**Comment by cheit-epic** [comment]
I want to push back on this again. How exactly would native JS decorators behave this way? In the example above, `foobar` isn't even decorated, so why should the presence of a decorator affect it?

If you instead had the below code, then it _would_ log 20. It feels very arbitrary that the order of the field in the class would change the runtime behavior.
```
export function decorator(target: any, context: any) {}

class Foo {
  something: number = 2; ///////////
  foobar: number;

  @decorator
  public bar() {}
}

Object.assign(Foo.prototype, { foobar: 20 });
console.log(new Foo().foobar);
```

As here, the emitted JS would become
```
constructor() {
  this.something = (__runInitializers(this, _instanceExtraInitializers), 2);
  //foobar is not even present in the constructor
}
```

**Comment by zivarah** [comment]
@fatcerberus in light of https://github.com/microsoft/TypeScript/issues/54264#issuecomment-1595352559, can this get re-opened? I think this issue is not explained by vanilla JS decorator behavior, but is instead a bug in the transpilation process. A declared member should consistently either be included or not included in the transpiled JavaScript regardless of the presence of decorators being used on other arbitrary members in the same class.

More explicit example of this particular behavior:

```typescript
export class MyClass
{
	public foo: () => void;

	public bar() { }
}
```

transpiles to:
```javascript
    var MyClass = /** @class */ (function () {
        function MyClass() {
        }
        MyClass.prototype.bar = function () { };
        return MyClass;
    }());
    exports.MyClass = MyClass;
```

whereas:

```typescript
export class MyClass
{
	public foo: () => void;

	@decorator
	public bar() { }
}
```

transpiles to:

```javascript
    var MyClass = exports.MyClass = function () {
        var _a;
        var _instanceExtraInitializers_1 = [];
        var _bar_decorators;
        return _a = /** @class */ (function () {
                function MyClass() {
                    this.foo = (__runInitializers(this, _instanceExtraInitializers_1), void 0);
                }
                MyClass.prototype.bar = function () { };
                return MyClass;
            }()),
            (function () {
                _bar_decorators = [decorator];
                __esDecorate(_a, null, _bar_decorators, { kind: "method", name: "bar", static: false, private: false, access: { has: function (obj) { return "bar" in obj; }, get: function (obj) { return obj.bar; } } }, null, _instanceExtraInitializers_1);
            })(),
            _a;
    }();
```

Where `foo` is not included in the transpiled javascript at all when the decorator is not applied.

Perhaps I'm just misunderstanding something here, but this seems like an undesired transpilation difference.

This ends up being a huge gotcha for scenarios where a prototype may be modified at runtime.

**Comment by fatcerberus** [comment]
I'm not a maintainer - I can't open and close issues.

Anyway, using a decorator causes TS to selectively enable `useDefineForClassFields` for the class, to ensure the class is spec-compliant.  This has the side effect that uninitialized properties will make it into the emit, because that's what the spec requires ([[Define]] semantics).

To be 100% clear, when I said:

> This is the expected behavior; it’s what this code would do in vanilla JS with native decorators.

...what I mean is that, if you were to run the OP code as written in an engine that natively supports decorators (and therefore native class fields), you'd necessarily observe the same behavior (printing `undefined`) as in the OP.  The non-`useDefineForClassFields` behavior of TS is effectively legacy behavior and shouldn't be assumed to play well with new ES features.

tl;dr: Adding a decorator to a class forces the whole class into spec-compliance mode.

**Comment by zivarah** [comment]
Thanks for the additional info.

First off - forgive my ignorance, this is all very unfamiliar to me and I'm just doing my best to understand the details.

> This has the side effect that uninitialized properties will make it into the emit

This doesn't appear to be true, or at least not that simply. Let's add another undefined property:

```typescript
export class MyClass
{
	public foo: () => void;
	public bar: () => void;

	@decorator
	public baz() { }
}
```

This transpiles to: 
```javascript
    var MyClass = exports.MyClass = function () {
        var _a;
        var _instanceExtraInitializers_1 = [];
        var _baz_decorators;
        return _a = /** @class */ (function () {
                function MyClass() {
                    this.foo = (__runInitializers(this, _instanceExtraInitializers_1), void 0);
                }
                MyClass.prototype.baz = function () { };
                return MyClass;
            }()),
            (function () {
                _baz_decorators = [decorator];
                __esDecorate(_a, null, _baz_decorators, { kind: "method", name: "baz", static: false, private: false, access: { has: function (obj) { return "baz" in obj; }, get: function (obj) { return obj.baz; } } }, null, _instanceExtraInitializers_1);
            })(),
            _a;
    }();
```

Notably, even though `foo` and `bar` are defined identically, _only_ `foo` ends up in the emit, simply because it comes first.

This is why it feels like a transpilation bug rather than expected behavior: Only the first field is impacted, (seemingly) arbitrarily.

**Comment by fatcerberus** [comment]
Huh, that *is* weird.

**Comment by zivarah** [comment]
> Huh, that is weird.

Agreed 😄

@DanielRosenwasser are you the right person to reconsider and potentially re-open this issue in light of this new detail?
